File: frontend/src/graphql/thesaurus.ts
import { gql } from '@apollo/client/core'

// =========================================================================
// Thesaurus Queries (Requirements 3.1, 3.2, 3.4, 9.1)
// =========================================================================

/**
 * Query paginated term pairs with optional filtering
 */
export const TERM_PAIRS_QUERY = gql`
  query TermPairs(
    $languagePairId: String!
    $catalogId: String
    $search: String
    $page: Int
    $pageSize: Int
  ) {
    termPairs(
      languagePairId: $languagePairId
      catalogId: $catalogId
      search: $search
      page: $page
      pageSize: $pageSize
    ) {
      items {
        id
        languagePairId
        catalogId
        sourceTerm
        targetTerm
        createdAt
        updatedAt
      }
      total
      page
      pageSize
      hasNext
    }
  }
`

/**
 * Query all catalogs for a language pair with term counts
 */
export const CATALOGS_QUERY = gql`
  query Catalogs($languagePairId: String!) {
    catalogs(languagePairId: $languagePairId) {
      id
      languagePairId
      name
      description
      termCount
      createdAt
      updatedAt
    }
  }
`

/**
 * Export term pairs as CSV content
 */
export const EXPORT_TERMS_CSV_QUERY = gql`
  query ExportTermsCsv($languagePairId: String!, $catalogId: String!) {
    exportTermsCsv(languagePairId: $languagePairId, catalogId: $catalogId)
  }
`

// =========================================================================
// Thesaurus Mutations (Requirements 1.1, 1.3, 1.5, 2.1, 4.1, 4.2, 5.1-5.4)
// =========================================================================

/**
 * Add or update a term pair (upsert behavior)
 */
export const ADD_TERM_PAIR_MUTATION = gql`
  mutation AddTermPair(
    $languagePairId: String!
    $catalogId: String!
    $sourceTerm: String!
    $targetTerm: String!
  ) {
    addTermPair(
      languagePairId: $languagePairId
      catalogId: $catalogId
      sourceTerm: $sourceTerm
      targetTerm: $targetTerm
    ) {
      id
      languagePairId
      catalogId
      sourceTerm
      targetTerm
      createdAt
      updatedAt
    }
  }
`

/**
 * Edit an existing term pair's target term
 */
export const EDIT_TERM_PAIR_MUTATION = gql`
  mutation EditTermPair($termId: String!, $targetTerm: String!) {
    editTermPair(termId: $termId, targetTerm: $targetTerm) {
      id
      languagePairId
      catalogId
      sourceTerm
      targetTerm
      createdAt
      updatedAt
    }
  }
`

/**
 * Delete a term pair by ID
 */
export const DELETE_TERM_PAIR_MUTATION = gql`
  mutation DeleteTermPair($termId: String!) {
    deleteTermPair(termId: $termId)
  }
`

/**
 * Delete all term pairs in a catalog
 */
export const BULK_DELETE_TERM_PAIRS_MUTATION = gql`
  mutation BulkDeleteTermPairs($languagePairId: String!, $catalogId: String!) {
    bulkDeleteTermPairs(languagePairId: $languagePairId, catalogId: $catalogId)
  }
`

/**
 * Import term pairs from CSV content
 */
export const IMPORT_TERMS_CSV_MUTATION = gql`
  mutation ImportTermsCsv(
    $languagePairId: String!
    $catalogId: String!
    $csvContent: String!
  ) {
    importTermsCsv(
      languagePairId: $languagePairId
      catalogId: $catalogId
      csvContent: $csvContent
    ) {
      created
      updated
      skipped
      errors
    }
  }
`

/**
 * Create a new catalog
 */
export const CREATE_CATALOG_MUTATION = gql`
  mutation CreateCatalog(
    $languagePairId: String!
    $name: String!
    $description: String
  ) {
    createCatalog(
      languagePairId: $languagePairId
      name: $name
      description: $description
    ) {
      id
      languagePairId
      name
      description
      termCount
      createdAt
      updatedAt
    }
  }
`

/**
 * Update a catalog's name or description
 */
export const UPDATE_CATALOG_MUTATION = gql`
  mutation UpdateCatalog(
    $languagePairId: String!
    $catalogId: String!
    $name: String
    $description: String
  ) {
    updateCatalog(
      languagePairId: $languagePairId
      catalogId: $catalogId
      name: $name
      description: $description
    ) {
      id
      languagePairId
      name
      description
      termCount
      createdAt
      updatedAt
    }
  }
`

/**
 * Delete a catalog and all its term pairs
 */
export const DELETE_CATALOG_MUTATION = gql`
  mutation DeleteCatalog($languagePairId: String!, $catalogId: String!) {
    deleteCatalog(languagePairId: $languagePairId, catalogId: $catalogId)
  }
`

File: backend/src/models/thesaurus.py
"""
Thesaurus data models for term translation pairs and catalogs.

This module defines the data structures used for managing term translation
pairs organized by language pair and catalog.
"""
from dataclasses import dataclass, field
from datetime import datetime, timezone
from typing import List, Optional


@dataclass
class TermPair:
    """
    A term translation pair mapping a source term to its target translation.
    
    Attributes:
        id: Unique identifier (UUID)
        language_pair_id: Reference to language pair (e.g., "zh-vi")
        catalog_id: Reference to catalog
        source_term: Source language term
        target_term: Target language translation
        created_at: Creation timestamp
        updated_at: Last modification timestamp
    """
    id: str
    language_pair_id: str
    catalog_id: str
    source_term: str
    target_term: str
    created_at: datetime = field(default_factory=lambda: datetime.now(timezone.utc))
    updated_at: datetime = field(default_factory=lambda: datetime.now(timezone.utc))
    
    def to_dict(self) -> dict:
        """Convert to dictionary for storage."""
        return {
            "id": self.id,
            "language_pair_id": self.language_pair_id,
            "catalog_id": self.catalog_id,
            "source_term": self.source_term,
            "target_term": self.target_term,
            "created_at": self.created_at.isoformat() if isinstance(self.created_at, datetime) else self.created_at,
            "updated_at": self.updated_at.isoformat() if isinstance(self.updated_at, datetime) else self.updated_at,
        }
    
    @classmethod
    def from_dict(cls, data: dict) -> "TermPair":
        """Create TermPair from dictionary."""
        created_at = data.get("created_at")
        updated_at = data.get("updated_at")
        
        # Parse datetime strings if needed
        if isinstance(created_at, str):
            created_at = datetime.fromisoformat(created_at.replace("Z", "+00:00"))
        if isinstance(updated_at, str):
            updated_at = datetime.fromisoformat(updated_at.replace("Z", "+00:00"))
        
        return cls(
            id=data["id"],
            language_pair_id=data["language_pair_id"],
            catalog_id=data["catalog_id"],
            source_term=data["source_term"],
            target_term=data["target_term"],
            created_at=created_at or datetime.now(timezone.utc),
            updated_at=updated_at or datetime.now(timezone.utc),
        )


@dataclass
class Catalog:
    """
    A catalog for organizing term pairs by domain or project.
    
    Attributes:
        id: Unique identifier (UUID)
        language_pair_id: Reference to language pair
        name: Catalog name
        description: Optional description
        created_at: Creation timestamp
        updated_at: Last modification timestamp
    """
    id: str
    language_pair_id: str
    name: str
    description: Optional[str] = None
    created_at: datetime = field(default_factory=lambda: datetime.now(timezone.utc))
    updated_at: datetime = field(default_factory=lambda: datetime.now(timezone.utc))
    
    def to_dict(self) -> dict:
        """Convert to dictionary for storage."""
        return {
            "id": self.id,
            "language_pair_id": self.language_pair_id,
            "name": self.name,
            "description": self.description or "",
            "created_at": self.created_at.isoformat() if isinstance(self.created_at, datetime) else self.created_at,
            "updated_at": self.updated_at.isoformat() if isinstance(self.updated_at, datetime) else self.updated_at,
        }
    
    @classmethod
    def from_dict(cls, data: dict) -> "Catalog":
        """Create Catalog from dictionary."""
        created_at = data.get("created_at")
        updated_at = data.get("updated_at")
        
        # Parse datetime strings if needed
        if isinstance(created_at, str):
            created_at = datetime.fromisoformat(created_at.replace("Z", "+00:00"))
        if isinstance(updated_at, str):
            updated_at = datetime.fromisoformat(updated_at.replace("Z", "+00:00"))
        
        return cls(
            id=data["id"],
            language_pair_id=data["language_pair_id"],
            name=data["name"],
            description=data.get("description"),
            created_at=created_at or datetime.now(timezone.utc),
            updated_at=updated_at or datetime.now(timezone.utc),
        )


@dataclass
class CatalogWithCount(Catalog):
    """
    Catalog with term pair count for display purposes.
    
    Attributes:
        term_count: Number of term pairs in the catalog
    """
    term_count: int = 0
    
    def to_dict(self) -> dict:
        """Convert to dictionary for storage."""
        result = super().to_dict()
        result["term_count"] = self.term_count
        return result
    
    @classmethod
    def from_catalog(cls, catalog: Catalog, term_count: int) -> "CatalogWithCount":
        """Create CatalogWithCount from a Catalog and term count."""
        return cls(
            id=catalog.id,
            language_pair_id=catalog.language_pair_id,
            name=catalog.name,
            description=catalog.description,
            created_at=catalog.created_at,
            updated_at=catalog.updated_at,
            term_count=term_count,
        )


@dataclass
class ImportResult:
    """
    Result of a CSV import operation.
    
    Attributes:
        created: Number of new term pairs created
        updated: Number of existing term pairs updated
        skipped: Number of invalid rows skipped
        errors: List of error messages for skipped rows
    """
    created: int = 0
    updated: int = 0
    skipped: int = 0
    errors: List[str] = field(default_factory=list)
    
    @property
    def total_processed(self) -> int:
        """Total number of rows processed."""
        return self.created + self.updated + self.skipped


@dataclass
class PaginatedTermPairs:
    """
    Paginated list of term pairs.
    
    Attributes:
        items: Term pairs for current page
        total: Total number of matching term pairs
        page: Current page number (1-indexed)
        page_size: Items per page
        has_next: Whether there are more pages
    """
    items: List[TermPair]
    total: int
    page: int
    page_size: int
    has_next: bool
    
    @property
    def total_pages(self) -> int:
        """Calculate total number of pages."""
        if self.page_size <= 0:
            return 0
        return (self.total + self.page_size - 1) // self.page_size
